File: chroma-league/src/main/java/net/booone/chromaleague/state/GameStateHelper.java
package net.booone.chromaleague.state;

import net.booone.chromaleague.hud.parts.GoldBar;
import net.booone.chromaleague.rest.activeplayer.ChampionStats;
import net.booone.chromaleague.rest.eventdata.DragonType;
import net.booone.chromaleague.rest.gamestats.GameStats;
import net.booone.chromaleague.tasks.FetchGameStatsTask;

import java.util.List;
import java.util.Objects;

public final class GameStateHelper {

    public static final long FIRST_ELDER_TIME = 150L;
    public static final long NEXT_ELDER_TIME = 300L;
    public static final long BARON_TIME = 180L;

    private GameStateHelper() {
    }

    public static boolean isActivePlayerAlive() {
        return !RunningState.getGameState().getPlayerList().getActivePlayer().isDead();
    }

    public static int getHpPercentage() {
        final ChampionStats championStats = RunningState.getGameState().getActivePlayer().championStats();
        return getPercentage(championStats.currentHealth(), championStats.maxHealth());
    }

    public static int getResourcePercentage() {
        final ChampionStats championStats = RunningState.getGameState().getActivePlayer().championStats();
        return getPercentage(championStats.resourceValue(), championStats.resourceMax());
    }

    public static double getGold() {
        return RunningState.getGameState().getActivePlayer().currentGold();
    }

    public static int getLevel() {
        return RunningState.getGameState().getActivePlayer().level();
    }

    /**
     * Future's market rune can make percentage negative
     */
    public static int getGoldPercentage() {
        final int percentage = getPercentage(RunningState.getGameState().getActivePlayer().currentGold(), GoldBar.GOLD_FULL);
        return Math.max(percentage, 0);
    }

    public static double getActivePlayerRange() {
        return RunningState.getGameState().getActivePlayer().championStats().attackRange();
    }

    private static int getPercentage(double firstDouble, double secondDouble) {
        return Double.valueOf(firstDouble * 100 / secondDouble).intValue();
    }

    public static void startBaronBuff(double eventTime, double currentTimeForReconnection) {
        if (playerAliveAtEventTime(eventTime, currentTimeForReconnection)) {
            if (hasExpired(eventTime, currentTimeForReconnection, BARON_TIME)) {
                RunningState.getGameState().getEventData().setBaronBuffEnd(null);
            } else {
                RunningState.getGameState().getEventData().setBaronBuffEnd(eventTime + BARON_TIME);
            }
        }
    }

    private static boolean hasExpired(double eventTime, double currentTimeForReconnection, long expirationTime) {
        if (currentTimeForReconnection == 0d) {
            double gameTime = RunningState.getGameState().getGameStats().gameTime();
            return gameTime - eventTime > expirationTime;
        }
        return currentTimeForReconnection - eventTime > expirationTime;
    }

    public static boolean hasBaronBuff() {
        final EventData eventData = RunningState.getGameState().getEventData();
        GameStats gameStats = RunningState.getGameState().getGameStats();
        if (gameStats != null && eventData.getBaronBuffEnd() != null && isActivePlayerAlive()) {
            return gameStats.gameTime() < eventData.getBaronBuffEnd();
        }
        eventData.setBaronBuffEnd(null);
        return false;
    }

    public static void startElderBuff(double eventTime, double currentTimeForReconnection) {
        addKilledElder();
        if (playerAliveAtEventTime(eventTime, currentTimeForReconnection)) {
            final int totalEldersKilled = getTotalEldersKilled();
            if (totalEldersKilled == 1) {
                computeElderBuffTimes(eventTime, currentTimeForReconnection, FIRST_ELDER_TIME);
            } else if (totalEldersKilled > 1) {
                computeElderBuffTimes(eventTime, currentTimeForReconnection, NEXT_ELDER_TIME);
            }
        }
    }

    private static void computeElderBuffTimes(double eventTime, double currentTimeForReconnection, long elderBuffTime) {
        if (hasExpired(eventTime, currentTimeForReconnection, elderBuffTime)) {
            RunningState.getGameState().getEventData().setElderBuffEnd(null);
        } else {
            RunningState.getGameState().getEventData().setElderBuffEnd(eventTime + elderBuffTime);
        }
    }

    private static boolean playerAliveAtEventTime(double eventTime, double currentTimeOrReconnectionTime) {
        final EventData eventData = RunningState.getGameState().getEventData();
        double gameTime = RunningState.getGameState().getGameStats().gameTime();
        if (eventData.getDeathTime() == null) {
            return true;
        }
        double eventLocalTime = gameTime - (currentTimeOrReconnectionTime - eventTime);
        return !(eventLocalTime > eventData.getDeathTime() && eventLocalTime < eventData.getRespawnTime());
    }

    public static boolean hasElderBuff() {
        final EventData eventData = RunningState.getGameState().getEventData();
        GameStats gameStats = RunningState.getGameState().getGameStats();
        if (gameStats != null && eventData.getElderBuffEnd() != null && isActivePlayerAlive()) {
            return gameStats.gameTime() < eventData.getElderBuffEnd();
        }
        eventData.setElderBuffEnd(null);
        return false;
    }

    public static List<DragonType> getKilledDragons() {
        return RunningState.getGameState().getEventData().getKilledDragons();
    }

    public static int getTotalEldersKilled() {
        return RunningState.getGameState().getEventData().getTotalEldersKilled();
    }

    public static void addKilledElder() {
        final EventData eventData = RunningState.getGameState().getEventData();
        eventData.setTotalEldersKilled(eventData.getTotalEldersKilled() + 1);
    }

    public static int getActivePlayerKillingSpree() {
        return RunningState.getGameState().getEventData().getActivePlayerKillingSpree();
    }

    public static void addPlayerKill() {
        RunningState.getGameState().getEventData().setActivePlayerKillingSpree(
                RunningState.getGameState().getEventData().getActivePlayerKillingSpree() + 1);
    }

    public static int getActivePlayerAssistSpree() {
        return RunningState.getGameState().getEventData().getActivePlayerAssistSpree();
    }

    public static void addPlayerAssist() {
        RunningState.getGameState().getEventData().setActivePlayerAssistSpree(
                RunningState.getGameState().getEventData().getActivePlayerAssistSpree() + 1);
    }

    public static boolean shouldPlayRespawnAnimation() {
        final EventData eventData = RunningState.getGameState().getEventData();
        double respawnTimer = RunningState.getGameState().getPlayerList().getActivePlayer().respawnTimer();
        if (respawnTimer > 0.0 && respawnTimer < 3.0) {
            if (eventData.getRespawnIndicator() == RespawnIndicator.IDLE) {
                eventData.setRespawnIndicator(RespawnIndicator.CHARGING);
                return true;
            }
        } else if (eventData.getRespawnIndicator() == RespawnIndicator.CHARGING) {
            eventData.setRespawnIndicator(RespawnIndicator.IDLE);
        }
        return false;
    }

    public static boolean shouldPlayRiftAnimation() {
        EventData eventData = RunningState.getGameState().getEventData();
        if (Objects.equals(FetchGameStatsTask.DEFAULT_MAP_TERRAIN, RunningState.getGameState().getGameStats().mapTerrain())
                || eventData.didRiftAnimationPlay()) {
            return false;
        }
        eventData.setRiftAnimationPlayed(true);
        return true;
    }

}

File: src/state/GameStateHelper.ts
import {DragonType} from '../rest/eventdata/DragonType.js';
import {RespawnIndicator} from './RespawnIndicator.js';
import {RunningState} from './RunningState.js';

/** Duration of the first Elder Dragon buff in seconds. Source: League of Legends wiki */
export const FIRST_ELDER_TIME = 150;
/** Duration of subsequent Elder Dragon buffs in seconds. Source: League of Legends wiki */
export const NEXT_ELDER_TIME = 300;
/** Duration of the Baron Nashor buff in seconds. Source: League of Legends wiki */
export const BARON_TIME = 180;
export const DEFAULT_MAP_TERRAIN = 'Default';

export function isActivePlayerAlive(): boolean {
    const { playerList, playerRiotId } = RunningState.getGameState();
    if (!playerList || !playerRiotId) return true;
    return !playerList.getActivePlayer(playerRiotId).isDead;
}

export function getHpPercentage(): number {
    const stats = RunningState.getGameState().activePlayer?.championStats;
    if (!stats) return 0;
    return getPercentage(stats.currentHealth, stats.maxHealth);
}

export function getResourcePercentage(): number {
    const stats = RunningState.getGameState().activePlayer?.championStats;
    if (!stats) return 0;
    return getPercentage(stats.resourceValue, stats.resourceMax);
}

export function getGold(): number {
    return RunningState.getGameState().activePlayer?.currentGold ?? 0;
}

export function getLevel(): number {
    return RunningState.getGameState().activePlayer?.level ?? 1;
}

export function getGoldPercentage(goldFull: number): number {
    return Math.max(getPercentage(RunningState.getGameState().activePlayer?.currentGold ?? 0, goldFull), 0);
}

export function getActivePlayerRange(): number {
    return RunningState.getGameState().activePlayer?.championStats.attackRange ?? 0;
}

function getPercentage(a: number, b: number): number {
    return Math.trunc(a * 100 / b);
}

export function startBaronBuff(eventTime: number, currentTimeForReconnection: number): void {
    if (playerAliveAtEventTime(eventTime, currentTimeForReconnection)) {
        if (hasExpired(eventTime, currentTimeForReconnection, BARON_TIME)) {
            RunningState.getGameState().eventData.baronBuffEnd = null;
        } else {
            RunningState.getGameState().eventData.baronBuffEnd = eventTime + BARON_TIME;
        }
    }
}

function hasExpired(eventTime: number, currentTimeForReconnection: number, expirationTime: number): boolean {
    if (currentTimeForReconnection === 0) {
        const gameStats = RunningState.getGameState().gameStats;
        if (!gameStats) return false;
        return gameStats.gameTime - eventTime > expirationTime;
    }
    return currentTimeForReconnection - eventTime > expirationTime;
}

export function hasBaronBuff(): boolean {
    const eventData = RunningState.getGameState().eventData;
    const gameStats = RunningState.getGameState().gameStats;
    if (gameStats && eventData.baronBuffEnd !== null && isActivePlayerAlive()) {
        return gameStats.gameTime < eventData.baronBuffEnd;
    }
    eventData.baronBuffEnd = null;
    return false;
}

export function startElderBuff(eventTime: number, currentTimeForReconnection: number): void {
    addKilledElder();
    if (playerAliveAtEventTime(eventTime, currentTimeForReconnection)) {
        const total = getTotalEldersKilled();
        if (total === 1) {
            computeElderBuffTimes(eventTime, currentTimeForReconnection, FIRST_ELDER_TIME);
        } else if (total > 1) {
            computeElderBuffTimes(eventTime, currentTimeForReconnection, NEXT_ELDER_TIME);
        }
    }
}

function computeElderBuffTimes(eventTime: number, currentTimeForReconnection: number, elderBuffTime: number): void {
    if (hasExpired(eventTime, currentTimeForReconnection, elderBuffTime)) {
        RunningState.getGameState().eventData.elderBuffEnd = null;
    } else {
        RunningState.getGameState().eventData.elderBuffEnd = eventTime + elderBuffTime;
    }
}

function playerAliveAtEventTime(eventTime: number, currentTimeOrReconnectionTime: number): boolean {
    const eventData = RunningState.getGameState().eventData;
    const gameStats = RunningState.getGameState().gameStats;
    if (!gameStats) return true;
    const gameTime = gameStats.gameTime;
    if (eventData.deathTime === null) return true;
    const eventLocalTime = gameTime - (currentTimeOrReconnectionTime - eventTime);
    return !(eventLocalTime > eventData.deathTime && eventLocalTime < eventData.respawnTime!);
}

export function hasElderBuff(): boolean {
    const eventData = RunningState.getGameState().eventData;
    const gameStats = RunningState.getGameState().gameStats;
    if (gameStats && eventData.elderBuffEnd !== null && isActivePlayerAlive()) {
        return gameStats.gameTime < eventData.elderBuffEnd;
    }
    eventData.elderBuffEnd = null;
    return false;
}

export function getKilledDragons(): readonly DragonType[] {
    return RunningState.getGameState().eventData.getKilledDragons();
}

export function getTotalEldersKilled(): number {
    return RunningState.getGameState().eventData.totalEldersKilled;
}

export function addKilledElder(): void {
    RunningState.getGameState().eventData.totalEldersKilled++;
}

export function getActivePlayerKillingSpree(): number {
    return RunningState.getGameState().eventData.activePlayerKillingSpree;
}

export function addPlayerKill(): void {
    RunningState.getGameState().eventData.activePlayerKillingSpree++;
}

export function getActivePlayerAssistSpree(): number {
    return RunningState.getGameState().eventData.activePlayerAssistSpree;
}

export function addPlayerAssist(): void {
    RunningState.getGameState().eventData.activePlayerAssistSpree++;
}

export function shouldPlayRespawnAnimation(): boolean {
    const eventData = RunningState.getGameState().eventData;
    const { playerList, playerRiotId } = RunningState.getGameState();
    if (!playerList || !playerRiotId) return false;
    const respawnTimer = playerList.getActivePlayer(playerRiotId).respawnTimer;
    if (respawnTimer > 0.0 && respawnTimer < 3.0) {
        if (eventData.respawnIndicator === RespawnIndicator.IDLE) {
            eventData.respawnIndicator = RespawnIndicator.CHARGING;
            return true;
        }
    } else if (eventData.respawnIndicator === RespawnIndicator.CHARGING) {
        eventData.respawnIndicator = RespawnIndicator.IDLE;
    }
    return false;
}

export function shouldPlayRiftAnimation(): boolean {
    const eventData = RunningState.getGameState().eventData;
    if (eventData.riftAnimationPlayed) {
        return false;
    }
    const gameStats = RunningState.getGameState().gameStats;
    if (!gameStats || gameStats.mapTerrain === DEFAULT_MAP_TERRAIN) {
        return false;
    }
    eventData.riftAnimationPlayed = true;
    return true;
}
